File: algorithms/kmp/kmp.js
/**
 * Knutt Morris Pratt
 * pattern matching O(n) time
 * @param { String } pattern
 * @param { String } str
 * @return { Number } index of match or -1
 */
function kmp(str, pattern) {
  if (!str || !pattern) return -1;
  const lsp = new Array(pattern.length).fill(0);
  let j = 0;
  let i = 1;
  while (i < pattern.length) {
    if (pattern[i] === pattern[j]) {
      lsp[i] = j + 1;
      j++;
      i++;
    } else if (j > 0) {
      j = lsp[j - 1];
    } else {
      i++;
    }
  }

  (i = 0), (j = 0);
  while (i < str.length) {
    if (str[i] === pattern[j]) {
      i++;
      j++;
      if (j === pattern.length) return i - j;
    } else if (j > 0) {
      j = lsp[j - 1];
    } else {
      i++;
    }
  }
  return -1;
}

console.assert(kmp('', 'aab') == -1, "kmp('', 'aab') == -1");
console.assert(kmp('abc', '') == -1, "kmp('abc', '') == -1");
console.assert(kmp('aabaacaa', 'aaa') == -1, "kmp('aabaacaa', 'aaa') == -1");
console.assert(kmp('a', 'aa') == -1, "kmp('a', 'aa') == -1");
console.assert(kmp('aa', 'a') == 0, "kmp('aa', 'a') == 0");
console.assert(kmp('abcabc', 'abc') == 0, "kmp('abcccc', 'abc') == 0");
console.assert(kmp('aaabbcabc', 'abc') == 6, "kmp('aaabbcabc', 'abc') == 6");
console.assert(kmp('aaaaaaaabaab', 'aab') == 6, "kmp('aaaaaaaab', 'aab') == 6");

File: algorithms/kmp/kmp.py
def kmp(text, pattern):
    """Return index of matching string of -1
    Knutt Morris Pratt algorithm
    pattern matching O(n) time
    >>> kmp('', 'aab')
    -1
    >>> kmp('abc', '')
    -1
    >>> kmp('aabaacaa', 'aaa')
    -1
    >>> kmp('a', 'aa')
    -1
    >>> kmp('aa', 'a')
    0
    >>> kmp('abcabc', 'abc')
    0
    >>> kmp('aaabbcabc', 'abc')
    6
    >>> kmp('aaaaaaaabaab', 'aab')
    6
    """
    if not len(text) or not len(pattern):
        return -1

    lsp = [0] * len(pattern)
    j = 0
    i = 1
    while (i < len(pattern)):
        if pattern[i] == pattern[j]:
            lsp[i] = j + 1
            i += 1
            j += 1
        elif j > 0:
            j = lsp[j - 1]
        else:
            i += 1

    i = 0
    j = 0

    while (i < len(text)):
        if text[i] == pattern[j]:
            i += 1
            j += 1
            if j == len(pattern):
                return i - j
        elif j > 0:
            j = lsp[j - 1]
        else:
            i += 1
    return -1


if __name__ == "__main__":
    import doctest
    doctest.testmod()
